app/Http/Controllers/ViewAssetsController.php

```diff
@@ -80,190 +82,116 @@ public function getRequestItem($itemType, $itemId = null)
-            $logaction->created_at = $data['requested_date'] = date("Y-m-d H:i:s");
-            $data['asset_type'] = 'hardware';
-            if ($user->location_id) {
-                $logaction->location_id = $user->location_id;
-            }
-            $logaction->target_id = $data['user_id'] = Auth::user()->id;
-            $logaction->target_type = User::class;
-            $log = $logaction->logaction('requested');
-
-            $data['requested_by'] = $user->present()->fullName();
-            $data['asset_name'] = $asset->present()->name();
-
-            $settings = Setting::getSettings();
-
-            if (($settings->alert_email!='')  && ($settings->alerts_enabled=='1') && (!config('app.lock_passwords'))) {
-                Mail::send('emails.asset-requested', $data, function ($m) use ($user, $settings) {
-                    $m->to(explode(',', $settings->alert_email), $settings->site_name);
-                    $m->replyTo(config('mail.reply_to.address'), config('mail.reply_to.name'));
-                    $m->subject(trans('mail.asset_requested'));
-                });
-            }
-
-            $asset->request();
-
-
-            if ($settings->slack_endpoint) {
 
+        $data['item'] = $asset;
+        $data['target'] =  Auth::user();
+        $data['item_quantity'] = 1;
+        $settings = Setting::getSettings();
 
-                $slack_settings = [
-                    'username' => $settings->botname,
-                    'channel' => $settings->slack_channel,
-                    'link_names' => true
-                ];
+        $logaction = new Actionlog();
+        $logaction->item_id = $data['asset_id'] = $asset->id;
+        $logaction->item_type = $data['item_type'] = Asset::class;
+        $logaction->created_at = $data['requested_date'] = date("Y-m-d H:i:s");
 
-                $client = new \Maknz\Slack\Client($settings->slack_endpoint, $slack_settings);
+        if ($user->location_id) {
+            $logaction->location_id = $user->location_id;
+        }
+        $logaction->target_id = $data['user_id'] = Auth::user()->id;
+        $logaction->target_type = User::class;
 
-                try {
-                        $client->attach([
-                            'color' => 'good',
-                            'fields' => [
-                                [
-                                    'title' => 'REQUESTED:',
-                                    'value' => class_basename(strtoupper($logaction->item_type)).' asset <'.url('/').'/hardware/'.$asset->id.'/view'.'|'.$asset->present()->name().'> requested by <'.url('/').'/hardware/'.$asset->id.'/view'.'|'.Auth::user()->present()->fullName().'>.'
-                                ]
 
-                            ]
-                        ])->send('Asset Requested');
+        // If it's already requested, cancel the request.
+        if ($asset->isRequestedBy(Auth::user())) {
+            $asset->cancelRequest();
```

Comment:
Again the cleanup you're getting by breaking out the notifications here is absolutely great.

But I worry about a few things here - first off, the `getRequestedAsset` - is this a `GET` request? I don't think it should be if it is.

Next, I think the cancellation action should be a completely separate action. Using it as a toggle could run into problems - for example, what happens if two techs both load up the requestable items screen at the same time, then try and request the same item for the same user? Wouldn't the later-submitted one cancel the first one? Better to have the later one fail because 'the item has already been requested for that user'. This would require a little more logic on the front end - to point correctly to the right method based on whether or not the item is already requested, but I think it is worth the slight addition in complexity.

Reply:
This cancellation is only by the user. Note the Auth::user() call.